File: hasekll/Tautology/Parser.hs
module Parser (parse) where

import Types
import Control.Monad.State

type ParserState = State [Token]

parse :: [Token] -> Prop
parse = evalState parse'

parse' :: ParserState Prop
parse' = do
    prop <- parseProp
    tk <- currentToken
    case tk of
      End -> return prop
      _   -> error "parser error"


parseProp :: ParserState Prop
parseProp = do
    exp <- parseExp
    tk   <- currentToken
    case tk of
      OpOr -> do
          advnaceToken
          prop <- parseProp
          return $ POr exp prop
      _ -> return exp


parseExp :: ParserState Prop
parseExp = do
    term <- parseTerm
    tk   <- currentToken
    case tk of
      OpAnd -> do
          advnaceToken
          exp <- parseExp
          return $ PAnd term exp
      _ -> return term


parseTerm :: ParserState Prop
parseTerm = do
    tk <- currentToken
    case tk of
      Var name -> advnaceToken >> (return $ PVar name)
      LeftParen -> do
          advnaceToken
          prop <- parseProp
          tk   <- currentToken
          case tk of
            RightParen -> advnaceToken >> (return prop)
            _          -> error "paren not match"
      OpNot -> do
          advnaceToken
          prop <- parseTerm
          return $ PNot prop


currentToken :: ParserState Token
currentToken = get >>= return . head

advnaceToken :: ParserState ()
advnaceToken = modify (drop 1)



{--
Prop = Exp | Exp OpOr Prop
Exp  = Term | Term OpAnd Exp
Term = Var | OpNot Term | LeftParen Prop RightParen
--}

File: c/tautology/parser.c
#include <stdio.h>
#include <stdlib.h>
#include <strings.h>
#include "types.h"
#include "tokenizer.h"
#include "parser.h"

struct expression *parse_term();
struct expression *parse_exp();
struct expression *parse_prop();
struct expression *new_expression();
void   destroy_expression(struct expression*);

struct expression *parse_term()
{
	struct expression *term;

	if (current_token == NULL) return NULL;
	switch (current_token->type) {
	case Var:
		term = new_expression();
		term->type = VarExp;
		term->u.var_id = current_token->u.var_id;
		advanceToken();
		return term;
	case LeftParen:
		advanceToken();
		term = parse_prop();
		if (term == NULL) {
			return NULL;
		}
		switch (current_token->type) {
		case RightParen:
			advanceToken();
			return term;
		default:
			return NULL;
		}
	case OpNot:
		advanceToken();
		term = new_expression();
		term->type = NotExp;
		term->u.right = NULL;
		term->u.left = parse_term();
		if (term->u.left == NULL) {
			free(term);
			return NULL;
		}
		return term;
	default:
		return NULL;
	}
}

struct expression *parse_exp()
{
	struct expression *term = parse_term();
	if (term == NULL) {
		return NULL;
	}

	if (current_token == NULL) return NULL;

	switch(current_token->type) {
	case OpAnd:
		advanceToken();
		struct expression *exp = new_expression();
		exp->type = AndExp;
		exp->u.left = term;
		exp->u.right = parse_exp();
		if (exp->u.right == NULL) {
			free(exp);
			return NULL;
		}
		return exp;
	default:
		return term;
	}
}

struct expression *parse_prop()
{
	struct expression *exp = parse_exp();
	if (exp == NULL) {
		return NULL;
	}
	if (current_token == NULL) return NULL;
	switch(current_token->type) {
	case OpOr:
		advanceToken();
		struct expression *prop = new_expression();
		prop->type = OrExp;
		prop->u.left = exp;
		prop->u.right = parse_prop();
		if (prop->u.right == NULL) {
			free(exp);
			return NULL;
		}
		return prop;
	default:
		return exp;
	}
}

struct expression *parse()
{
	init();
	advanceToken();
	struct expression *prop = parse_prop();

	if (prop == NULL) {
		return NULL;
	}
	if (current_token == NULL) return NULL;
	switch(current_token->type) {
	case End:
		return prop;
	default:
		return NULL;
	}
}

struct expression *new_expression()
{
	struct expression *exp = malloc(sizeof(struct expression));
	bzero(exp, sizeof(struct expression));
	return exp;
}

void destroy_expression(struct expression *exp)
{
	if (exp == NULL) return;
	if (exp->type != VarExp) {
		destroy_expression(exp->u.left);
		destroy_expression(exp->u.right);
	}
	free(exp);
}
